src/group_theory/exponent.lean

```diff
@@ -0,0 +1,167 @@
+  for all `g`.-/
+@[to_additive "A predicate on an additive monoid saying that there is a positive integer `n` such
+  that `n • g = 0` for all `g`."]
+def exponent_exists  := ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1
+
+/--The exponent of a group is the smallest positive integer `n` such that `g ^ n = 1` for all
+  `g ∈ G` if it exists, otherwise it is zero by convention.-/
+@[to_additive "The exponent of an additive group is the smallest positive integer `n` such that
+  `n • g = 0` for all `g ∈ G` if it exists, otherwise it is zero by convention."]
+noncomputable def exponent :=
+if h : exponent_exists G then nat.find h else 0
+
+@[to_additive exponent_nsmul_eq_zero]
+lemma pow_exponent_eq_one (g : G) : g ^ exponent G = 1 :=
+begin
+  by_cases exponent_exists G,
+  { simp_rw [exponent, dif_pos h],
+    exact (nat.find_spec h).2 g},
+  { simp_rw [exponent, dif_neg h, pow_zero] }
+end
+
+@[to_additive nsmul_eq_mod_exponent]
+lemma pow_eq_mod_exponent {n : ℕ} (g : G): g ^ n = g ^ (n % exponent G) :=
+calc g ^ n = g ^ (n % exponent G + exponent G * (n / exponent G)) : by rw [nat.mod_add_div]
+  ... = g ^ (n % exponent G) : by simp [pow_add, pow_mul, pow_exponent_eq_one]
+
+@[to_additive]
+lemma exponent_pos_of_exists (n : ℕ) (hpos : 0 < n)
+(hG : ∀ g : G, g ^ n = 1) : 0 < exponent G :=
+begin
+  have h : ∃ n, 0 < n ∧ ∀ g : G, g ^ n = 1 := ⟨n, hpos, hG⟩,
+  rw [exponent, dif_pos],
+  exact (nat.find_spec h).1,
+end
+
+@[to_additive]
+lemma exponent_min' (n : ℕ) (hpos : 0 < n) (hG : ∀ g : G, g ^ n = 1) :
+  exponent G ≤ n :=
+begin
+  rw [exponent, dif_pos],
+  { apply nat.find_min',
+    exact ⟨hpos, hG⟩ },
+  { exact ⟨n, hpos, hG⟩ },
+end
+
+@[to_additive]
+lemma exponent_min (m : ℕ) (hpos : 0 < m) (hm : m < exponent G) : ∃ g : G, g ^ m ≠ 1 :=
+begin
+  by_contradiction,
+  push_neg at h,
+  have hcon : exponent G ≤ m := exponent_min' G m hpos h,
+  linarith,
+end
+
+lemma exp_punit_eq_one : exponent punit = 1 :=
+begin
+  apply le_antisymm,
+  { apply exponent_min' _ _ nat.one_pos,
+    intro g,
+    refl },
+  { apply nat.succ_le_of_lt,
+    apply exponent_pos_of_exists _ 1 (nat.one_pos),
+    intro g,
+    refl },
+end
+
+@[to_additive add_order_dvd_exponent]
+lemma order_dvd_exponent (g : G) : (order_of g) ∣ exponent G :=
+order_of_dvd_of_pow_eq_one (pow_exponent_eq_one G g)
+
+@[to_additive]
+lemma exponent_dvd_of_forall_pow_eq_one (n : ℕ) (hpos : 0 < n) (hG : ∀ g : G, g ^ n = 1) :
+  exponent G ∣ n :=
+begin
+  apply nat.dvd_of_mod_eq_zero,
+  by_contradiction h,
+  have h₁ := nat.pos_of_ne_zero h,
+  have h₂ : n % exponent G < exponent G := nat.mod_lt _ (exponent_pos_of_exists _ n hpos hG),
+  have h₃ : exponent G ≤ n % exponent G,
+  { apply exponent_min' _ _ h₁,
+    simp_rw ←pow_eq_mod_exponent,
+    exact hG },
+  linarith,
+end
+
+variable [fintype G]
+
+@[to_additive lcm_add_order_of_dvd_exponent]
+lemma lcm_order_of_dvd_exponent : ((finset.univ : finset G).image order_of).lcm id ∣ exponent G :=
```

Comment:
oh, why are we stating this as `finset.univ.image order_of.lcm id` instead of `finset.univ.lcm order_of`? I think that's easier to work with and probably a bit easier to prove things with too

Reply:
Could you elaborate? Just pasting what you wrote instead of what I wrote doesn't work.

Reply:
Ah, I think I got it to work, let me try to change the remainder of the file to use it.

Reply:
Let me know if this is what you had in mind.

Reply:
Yes, it looks good!